ndb_api/time_machine.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022-2023 Nutanix, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ndb_api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/nutanix-cloud-native/ndb-operator/ndb_client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Takes a snapshot for a time machine
+// Returns the task info summary response for the operation TODO
+func TakeSnapshotForTM(ctx context.Context, ndbClient *ndb_client.NDBClient, tmName string) (task TaskInfoSummaryResponse, err error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Entered ndb_api.TakeSnapshotForTM")
+	if ndbClient == nil {
+		err = errors.New("nil reference")
+		log.Error(err, "Received nil ndbClient reference")
+		return
+	}
+	// Checking if id is empty, this is necessary to create a snapshot for the timemachine (/tms/{timemachine_id}/snapshots)
+	if tmName == "" {
+		err = fmt.Errorf("snapshot id is empty")
+		log.Error(err, "no snapshot id provided")
+		return
+	}
+	takeSnapshotPath := fmt.Sprintf("tms/%s/snapshots", tmName)
+	requestBody := &SnapshotRequest{
+		Name: "snapshot2",
+		SnapshotLcmConfig: SnapshotLcmConfig{
+			SnapshotLCMConfigDetailed: SnapshotLcmConfigDetailed{
+				ExpiryDetails: ExpiryDetails{
+					ExpiryDateTimezone: "Asia/Calcutta",
+					ExpireInDays:       "10",
+				},
+			},
+		},
+	}
```

Comment:
This is better suited in a helper file. Maybe create a function - GenerateSnapshotRequest?

Reply:
Ok makes sense, will do.